webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/hooks/RunHooks.scala

```diff
@@ -304,42 +323,77 @@ object RunHooks {
-     IOResult.effect {
-       val dir = new File(basePath)
-       // Check that dir exists before looking in it
-       if (dir.exists) {
-         HooksLogger.debug(s"Looking for hooks in directory '${basePath}', ignoring files with suffix: '${ignoreSuffixes.mkString("','")}'")
-         // only keep executable files
-         val files = dir.listFiles().toList.flatMap { file =>
-           file match {
-             case f if (f.isDirectory) => None
-             case f =>
-               if(f.canExecute) {
-                 val name = f.getName
-                 //compare ignore case (that's why it's a regienMatches) extension and name
-                 ignoreSuffixes.find(suffix => name.regionMatches(true, name.length - suffix.length, suffix, 0, suffix.length)) match {
-                   case Some(suffix) =>
-                     HooksLogger.debug(s"Ignoring hook '${f.getAbsolutePath}' because suffix '${suffix}' is in the ignore list")
-                     None
-                   case None      =>
-                     Some(f.getName)
-                 }
-               } else {
-                 HooksLogger.debug(s"Ignoring hook '${f.getAbsolutePath}' because it is not executable. Check permission if not expected behavior.")
-                 None
-               }
-           }
-         }.sorted // sort them alphanumericaly
-         Hooks(basePath, files)
-       } else {
-         HooksLogger.debug(s"Ignoring hook directory '${dir.getAbsolutePath}' because path does not exists")
-         // return an empty Hook
-         Hooks(basePath, List[String]())
-       }
-     }
-   }
+  def getHooks(basePath: String, ignoreSuffixes: List[String]): Box[Hooks] = getHooksPure(basePath, ignoreSuffixes).toBox
+
+  /*
+   * Check for the existence of hook limit in the 10 first lines of the file
+   * This method is effectful, needs to be wrap in an IO.
+   */
+  def effectfulGetHookTimeout(hook: File): HookTimeout = {
+    def get(filename: String, lines: List[String], varName: String): Option[Duration] = {
+      val pattern = s""".* ${varName}\\s*=\\s*(.+)""".r
+      val pf: PartialFunction[String, Duration] = ((line: String) => line match {
+        case pattern(timeout) =>
+          try {
+            Some(Duration.fromScala(scala.concurrent.duration.Duration(timeout)))
+          } catch {
+            case ex:NumberFormatException =>
+              HooksLogger.warn(s"Error: in '${filename}', can not parse ${varName} value as a duration. Expecting format: " +
+                               s"<length>\\s+[<short unit><long unit>], for ex: 1h, 3 minutes, etc: ${timeout}")
+              None
+          }
+        case _ => None
+      }).unlift
+
+      lines.collectFirst(pf)
+    }
+
+    val filename = hook.getAbsolutePath
+    val lines = {
+      try {
+        new String(better.files.File(filename).chars(StandardCharsets.UTF_8).take(1000).toArray).split("\n").toList
```

Comment:
I would not limit to 1000 characters, as people may want to have a lot of documentation at the start of a script. We ought to read the totallity of script if possible (how often is it read?)

Reply:
shouldn't it be a stream that we consume to find the line if possible ? So that we don't use memory ?

Reply:
well, most of the time, they won't be here. Perhaps it's of not use, I don't know (but it's already merged unfortunatly)

Reply:
The "already merged" part is false, that was an other PR :)
I think again to it, and I really think it's an OK limit: the need is rare, and an user who needs it will just read the doc and put the comment on top, before script doc, and it will be OK.

Reply:
ok

Reply:
hum, spliting on \n may fail on windows encoded file

Reply:
you should probably use file.lineIterator to get lines

Reply:
these are server side script, they need to be unix encoded. I con't use file.lineIterator, because the stream is not correctly closed if it is not totally consumed, which is the standard case here.